### ISSUE TITLE
Streaming version

### ISSUE THREAD

**Issue by xenia-kra** [huggingface/diffusers#3610]
Looking for a way to optimize training on a large dataset. I would like to be able to use load_dataset(streaming=True...)
From what I see, current version doesn't support it. How hard is to adapt the current code? Any other suggestions to deal with the problem?

**Comment by pcuenca** [comment]
Broadly, I think the steps would be:
- Add a new arg `--dataset_streaming`, of course.
- Make `--image_column` and `--caption_column` mandatory in this case, as we won't have access to the column names when we create the dataset.
- Provide an estimate for the total number of items in the dataset, as it's used to compute the number of epochs for the training loop. Another approach would be to change training so it's only based on steps not epochs.
- Don't shuffle the full dataset or the dataloader. Instead, use something like `dataset.shuffle(buffer_size=1000)` to shuffle collections of (in this case) 1k items.

This is assuming a single node, for distributed training with dataset sharding there might be additional complications. I'm not too familiar with that environment so I'll let others chime in on that regard.

May I ask if you can share what's your use case?

**Comment by sayakpaul** [comment]
An example dataset with streaming capabilities is available here:
https://github.com/huggingface/community-events/blob/d095130e498887b666e68e4e2875351e3ca9c8aa/jax-controlnet-sprint/training_scripts/train_controlnet_flax.py#L518

**Comment by pcuenca** [comment]
Ohh, nice!
